File: infrastructure/docker/Dockerfile.go
FROM golang:1.21-bullseye AS builder

# Install build dependencies
RUN apt-get update && apt-get install -y git gcc make && rm -rf /var/lib/apt/lists/*

WORKDIR /build

# Copy go mod files
COPY go.mod ./

# Copy source code
COPY . .

# Download dependencies and tidy
RUN go mod tidy && go mod download

# Build the binary
ARG SERVICE=prometheus-collector
RUN CGO_ENABLED=1 GOOS=linux go build -a \
    -ldflags="-w -s" -o /collector ./${SERVICE}/cmd

# Runtime image
FROM debian:bullseye-slim

RUN apt-get update && apt-get install -y --no-install-recommends \
    ca-certificates wget librdkafka1 \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /app

# Copy the binary
COPY --from=builder /collector ./collector

# Create non-root user
RUN useradd -r -u 1000 -s /bin/false collector
USER collector

# Health check
HEALTHCHECK --interval=30s --timeout=3s --start-period=10s --retries=3 \
  CMD wget --no-verbose --tries=1 --spider http://localhost:8080/health || exit 1

# Run the collector
ENTRYPOINT ["./collector"]

File: infrastructure/docker/Dockerfile.java
FROM openjdk:11-jdk-slim AS builder

WORKDIR /build

# Copy Gradle files
COPY build.gradle settings.gradle gradlew ./
COPY gradle ./gradle

# Download dependencies
RUN ./gradlew dependencies --no-daemon

# Copy source code
COPY src ./src

# Build the application
RUN ./gradlew shadowJar --no-daemon

# Runtime image
FROM openjdk:11-jre-slim

WORKDIR /app

# Copy the fat JAR
COPY --from=builder /build/build/libs/*.jar ./app.jar

# Create non-root user
RUN useradd -r -u 1000 flink
USER flink

# Health check
HEALTHCHECK --interval=30s --timeout=3s --start-period=60s --retries=3 \
  CMD test -f /tmp/flink-health || exit 1

# Run the Flink job
ENTRYPOINT ["java", "-jar", "app.jar"]
